modules/dnn/src/registry.cpp

```diff
@@ -121,7 +125,8 @@ class BackendRegistry
         if (haveCUDA())
         {
             backends.push_back(std::make_pair(DNN_BACKEND_CUDA, DNN_TARGET_CUDA));
-            backends.push_back(std::make_pair(DNN_BACKEND_CUDA, DNN_TARGET_CUDA_FP16));
+            if (cuda4dnn::doesDeviceSupportFP16())
+                backends.push_back(std::make_pair(DNN_BACKEND_CUDA, DNN_TARGET_CUDA_FP16));
```

Comment:
It works well with single gpu configuration. The current GPU may be old and does not support FP16, but the second one does. It's popular case, if one GPU is used for rendering another one - for compute.

Reply:
It is fine to check the **CURRENT** CUDA device.
Lets left the caller responsibility to properly select used device via `cudaSetDevice` before these calls.

BTW, OpenCL backend has similar problems but looks like they are handled in another place.